crates/uv-requirements/src/specification.rs

```diff
@@ -223,15 +227,84 @@ impl RequirementsSpecification {
         requirements: &[RequirementsSource],
         constraints: &[RequirementsSource],
         overrides: &[RequirementsSource],
+        groups: &[PipGroupName],
         client_builder: &BaseClientBuilder<'_>,
     ) -> Result<Self> {
         let mut spec = Self::default();
 
+        // Resolve sources into specifications so we know their `source_tree`s
+        let mut requirement_sources = Vec::new();
+        for source in requirements {
+            let source = Self::from_source(source, client_builder).await?;
+            requirement_sources.push(source);
+        }
+
+        // pip `--group` flags specify their own sources, which we need to process here
```

Comment:
I'm pretty sure this makes sense, but it might be relevant for Charlie to take a look at before merging.